### PR TITLE
Kozel translate to sc

### PR DESCRIPTION
**Pull request by PtaxaMagic**
Graph Translation with temporal decomposition.
still have problems with nonoriented edges.
No rename/create graph after the name is changed.

### CODE REVIEW

File: sc-web/components/scgg/src/scgg-struct.js [hunk 1]
```diff
@@ -233,19 +233,13 @@ function scggScStructTranslator(_editor, _sandbox) {
         updateFromSc: function(added, element, arc) {
             scggFromSc.update(added, element, arc);
         },
-        
-        translateToSc: function(callback) {
+
+translateToSc: function(callback) {
+                       //wtf
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: wtf

File: sc-web/components/scgg/src/scgg-struct.js [hunk 2]
```diff
@@ -233,19 +233,13 @@ function scggScStructTranslator(_editor, _sandbox) {
...
 
-            var dfdNodes = jQuery.Deferred();
-
-            editor.scene.commandManager.clear();
-            var nodes = editor.scene.nodes.slice();
-            var links = editor.scene.links.slice();
-            var buses = editor.scene.buses.slice();
-            var objects = [];
-            
+            var addrStruct = undefined;
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: var addrStruct = null; 
or 
var addrStruct;

File: sc-web/components/scgg/src/scgg-struct.js [hunk 3]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
-                            node.setObjectState(SCggObjectState.NewInMemory);
-                            objects.push(node);
-                            if (node.text) {
-                                translateGtIdentifier(node)
-                                    .done(dfd.resolve)
-                                    .fail(dfd.reject);
-                            } else {
-                                dfd.resolve();
+            var addrStruct;
+            // editor.scene.comandManager.clear();
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: uncomment

File: sc-web/components/scgg/src/scgg-struct.js [hunk 4]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                var dfd = new jQuery.Deferred();
+                var createNodeGraph = function() {
+                    var dfdAddrStruct = new jQuery.Deferred();
+
+                    window.sctpClient.create_node(sc_type_node | sc_type_const | sc_type_node_struct).done(function (nodeNewGraph){
+                        addrStruct = nodeNewGraph;
+
+                        window.sctpClient.create_arc(sc_type_arc_pos_const_perm, SCggKeynodesHandler.scKeynodes.concept_graph, nodeNewGraph)
+                        .done(dfdAddrStruct.resolve)
+                        .fail(dfdAddrStruct.reject);;
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: ;;

File: sc-web/components/scgg/src/scgg-struct.js [hunk 5]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+
+                    });
+                        return dfdAddrStruct.promise();
+                };
+
+                var translateTemporalDecomposition = function() {
+                    var dfdTranslateDecomposition = new jQuery.Deferred();
+                    var graphName = editor.getGraphName();
+                    var graphNameAddr = editor.render.sandbox.addr;
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: not use
graph name addr in editor.render.sandbox.graphNodeAddr

File: sc-web/components/scgg/src/scgg-struct.js [hunk 6]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+
+                    // var temporary_entity;
+                    
+                        var findDecomposition = function(){
+                            var dfdFind = new jQuery.Deferred();
+
+                            window.sctpClient.iterate_elements(SctpIteratorType.SCTP_ITERATOR_3A_A_F,
+                            [   sc_type_node | sc_type_const | sc_type_node_tuple,
+                                sc_type_arc_pos_const_perm,
+                                editor.render.sandbox.addr,
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: ,

File: sc-web/components/scgg/src/scgg-struct.js [hunk 7]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                    var graphNameAddr = editor.render.sandbox.addr;
+
+                    var graphDecomposition =  new SCgg.ModelNode({
+                                text: graphName
+                                // sc_addr: graphDecompositionAddr
+                            });
+
+                    // var temporary_entity;
+                    
+                        var findDecomposition = function(){
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: decomposition is find in SCggComponent
if not exist decomposition will be null

File: sc-web/components/scgg/src/scgg-struct.js [hunk 8]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                                    dfdFind.resolve();
+                                }).fail(dfdFind.reject);
+                            }).fail(dfdFind.reject);
+
+                            
+                            return dfdFind.promise();                        
+                        };
+
+                        var createDecomposition = function() {
+                            editor.render.sandbox.loadGraph = true; //?
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: yes

File: sc-web/components/scgg/src/scgg-struct.js [hunk 9]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+
+                            var dfdCreate = new jQuery.Deferred();
+
+                            window.sctpClient.create_node(sc_type_node | sc_type_const).done(function (graphDecompositionAddr) {
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: sc_type_node | sc_type_const | sc_type_node_tuple if this Decomposition

File: sc-web/components/scgg/src/scgg-struct.js [hunk 10]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                            
+                                graphDecomposition.sc_addr = graphDecompositionAddr;
+                                window.sctpClient.create_arc(sc_type_arc_pos_const_perm, SCggKeynodesHandler.scKeynodes.concept_graph, graphDecompositionAddr);
+                            
+                                    window.sctpClient.create_node(sc_type_node | sc_type_const | sc_type_node_tuple).done(function (graphDecompositionTupleAddr) {
+                                        editor.render.sandbox.decompositionNodeAddr = graphDecompositionTupleAddr;
+                                        editor.render.sandbox.graphNodeAddr = graphDecompositionAddr;
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: ...graphNodeAddr = graphDecompositionAddr;
wrong name, function or error?

File: sc-web/components/scgg/src/scgg-struct.js [hunk 11]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                    });
+                        return dfdAddrStruct.promise();
+                };
+
+                var translateTemporalDecomposition = function() {
+                    var dfdTranslateDecomposition = new jQuery.Deferred();
+                    var graphName = editor.getGraphName();
+                    var graphNameAddr = editor.render.sandbox.addr;
+
+                    var graphDecomposition =  new SCgg.ModelNode({
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: SCgg.ModelNode is model node in editor, decomposition not show in editor

File: sc-web/components/scgg/src/scgg-struct.js [hunk 12]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                        };
+
+                        var createDecomposition = function() {
+                            editor.render.sandbox.loadGraph = true; //?
+
+                            var dfdCreate = new jQuery.Deferred();
+
+                            window.sctpClient.create_node(sc_type_node | sc_type_const).done(function (graphDecompositionAddr) {
+                            
+                                graphDecomposition.sc_addr = graphDecompositionAddr;
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: not need

File: sc-web/components/scgg/src/scgg-struct.js [hunk 13]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                                window.sctpClient.create_arc(sc_type_arc_pos_const_perm, SCggKeynodesHandler.scKeynodes.concept_graph, graphDecompositionAddr);
+                            
+                                    window.sctpClient.create_node(sc_type_node | sc_type_const | sc_type_node_tuple).done(function (graphDecompositionTupleAddr) {
+                                        editor.render.sandbox.decompositionNodeAddr = graphDecompositionTupleAddr;
+                                        editor.render.sandbox.graphNodeAddr = graphDecompositionAddr;
+                                        window.sctpClient.create_arc(sc_type_arc_common | sc_type_const, graphDecompositionTupleAddr, graphDecomposition.sc_addr).done(function(arcTemporalDecomposition) {
+                                            window.sctpClient.create_arc(sc_type_arc_pos_const_perm, SCggKeynodesHandler.scKeynodes.nrel_temporal_decomposition, arcTemporalDecomposition).done(function(){
+                                                        if ((graphDecomposition.text !== '') && (graphDecomposition.text !== null)) {
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: check editor.getGraphName()

File: sc-web/components/scgg/src/scgg-struct.js [hunk 14]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                                    var dfdDeleteCurrent = new jQuery.Deferred();
+
+                                    window.sctpClient.iterate_elements(SctpIteratorType.SCTP_ITERATOR_5F_A_A_A_F,
+                                        [   editor.render.sandbox.decompositionNodeAddr,
+                                            sc_type_arc_pos_const_perm,
+                                            sc_type_node | sc_type_const,
+                                            sc_type_arc_pos_const_perm,
+                                            SCggKeynodesHandler.scKeynodes.rrel_current_version
+                                        ]).done(function(results){ 
+                                            SCWeb.core.Server.resolveScAddr(['sc_garbage'], function (keynodes) {
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: add sc_garbage in SCggKeynodesHandler

File: sc-web/components/scgg/src/scgg-struct.js [hunk 15]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                                                window.sctpClient.create_arc(sc_type_arc_pos_const_perm, keynodes.sc_garbage, results[0][3])
+                                                .done(dfdDeleteCurrent.resolve)
+                                                .fail(dfdDeleteCurrent.reject);
+                                            });
+                                            window.sctpClient.erase_element(results[0][2]).done(dfdDeleteCurrent.resolve)
+                                            .fail(dfdDeleteCurrent.reject);
+                                        }).fail(dfdDeleteCurrent.resolve);
+
+                                    return dfdDeleteCurrent.promise();
+                                }
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: ;

File: sc-web/components/scgg/src/scgg-struct.js [hunk 16]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+
+                                        window.sctpClient.create_arc(sc_type_arc_pos_const_perm, editor.render.sandbox.decompositionNodeAddr, addrStruct).done(function (arcSysIdtf){
+                                            window.sctpClient.create_arc(sc_type_arc_pos_const_perm, SCggKeynodesHandler.scKeynodes.rrel_current_version, arcSysIdtf)
+                                            .done(function(){
+                                                addCurrent.resolve();
+                                            }).fail(addCurrent.reject);        
+                                        }).fail(addCurrent.reject);
+
+                                    return addCurrent.promise();
+                                }
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: ;

File: sc-web/components/scgg/src/scgg-struct.js [hunk 17]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                    
 
-                    return dfd.promise();
-                }
-                
-                var funcs = [];
-                for (var i = 0; i < nodes.length; ++i) {
-                    funcs.push(fQueue.Func(implFunc, [ nodes[i] ]));
+                    return dfdTranslateDecomposition.promise();
                 }
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: ;

File: sc-web/components/scgg/src/scgg-struct.js [hunk 18]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                                    editor.render.sandbox.graphNodeAddr = results[0][2];
+                                    dfdFind.resolve();
+                                }).fail(dfdFind.reject);
+                            }).fail(dfdFind.reject);
+
+                            
+                            return dfdFind.promise();                        
+                        };
+
+                        var createDecomposition = function() {
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: may be createGraphAndDecomposition

File: sc-web/components/scgg/src/scgg-struct.js [hunk 19]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                           
+                       // findOrCreateDecomposition().done(dfdTranslateDecomposition.resolve);
+                        if (!editor.render.sandbox.loadGraph){
+                        createDecomposition().done(function() {
+                            addCurrentGraph().done(dfdTranslateDecomposition.resolve)
+                            .fail(dfdTranslateDecomposition.reject);
+                        }).fail(dfdTranslateDecomposition.reject);
+                       }
+                        else{
+                        if (editor.render.sandbox.graphNodeAddr === null){
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: if editor.render.sandbox.graphNodeAddr === null
then editor.render.sandbox.decompositionNodeAddr null to and you need create main graph node and decompositionNode

File: sc-web/components/scgg/src/scgg-struct.js [hunk 20]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                            }
+                            
+                            })
+                            //leave it for me right now :D
+                            .fail(function() {
+                                console.log("WTF THIS MISTAKE " + addrStruct + " " + nodeAddr )
+                            });
+                        });
+                        return dfdNode.promise();
                 }
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: ;

File: sc-web/components/scgg/src/scgg-struct.js [hunk 21]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
-                };
-                var funcs = [];
-                for (var i = 0; i < editor.scene.contours.length; ++i){
-                    editor.scene.contours[i].addNodesWhichAreInContourPolygon(editor.scene.nodes);
-                    editor.scene.contours[i].addNodesWhichAreInContourPolygon(editor.scene.links);
-                    editor.scene.contours[i].addEdgesWhichAreInContourPolygon(editor.scene.edges);
-                    funcs.push(fQueue.Func(scAddrGen, [ editor.scene.contours[i] ]));
+                            }
+                            
+                            })
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: need reformat code(alignment)

File: sc-web/components/scgg/src/scgg-struct.js [hunk 22]
```diff
@@ -256,112 +250,250 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                            //leave it for me right now :D
+                            .fail(function() {
+                                console.log("WTF THIS MISTAKE " + addrStruct + " " + nodeAddr )
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: can not create arc between addrStruct and nodeAddr

SCg do it yourself!!! check

File: sc-web/components/scgg/src/scgg-struct.js [hunk 23]
```diff
@@ -372,140 +504,100 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                        window.sctpClient.create_arc(sc_type_arc_common | sc_type_const, src, target).done(function(r){
+                            edge.setScAddr(r);
+                            window.sctpClient.create_arc(sc_type_arc_pos_const_perm, addrStruct, r).done(function (arcSystemIdentifier){
+                                window.sctpClient.create_arc(sc_type_arc_pos_const_perm, SCggKeynodesHandler.scKeynodes.rrel_oredge, arcSystemIdentifier)
+                                .done(dfdEdge.resolve(r));
+                            });
+                            
+                        });
+                        return dfdEdge.promise();
+                    }
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: ;

File: sc-web/components/scgg/src/scgg-struct.js [hunk 24]
```diff
@@ -372,140 +504,100 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                                    }).fail(dfdWeight.reject);
+                                }).fail(dfdWeight.reject);                                
+                        }
+                        else {
+                            dfdWeight.resolve();
+                        }
+
+
+                        return dfdWeight.promise();
+                    }
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: ;

File: sc-web/components/scgg/src/scgg-struct.js [hunk 25]
```diff
@@ -372,140 +504,100 @@ function scggScStructTranslator(_editor, _sandbox) {
...
+                                    objects.push(edge);
+                                    newxIteration();
+                                    });
+                                }) 
+                            });
+                        }
+                        else {
+                        createOrEdge(src,trg).done(function (r){
+                            translateWeight(edge);
                             edge.setObjectState(SCggObjectState.NewInMemory);
```
Comment (blrB) [inline sc-web/components/scgg/src/scgg-struct.js]: delete edge.setObjectState(SCggObjectState.NewInMemory);

File: sc-web/components/scgg/src/scgg-struct.js (hunk 7 above)
Comment (PtaxaMagic) [reply sc-web/components/scgg/src/scgg-struct.js]: if we open graph from struct. loadGraph will be true. and we have to find decomposition.

File: sc-web/components/scgg/src/scgg-struct.js (hunk 18 above)
Comment (PtaxaMagic) [reply sc-web/components/scgg/src/scgg-struct.js]: No. 
see the logic below

File: sc-web/components/scgg/src/scgg-struct.js (hunk 9 above)
Comment (PtaxaMagic) [reply sc-web/components/scgg/src/scgg-struct.js]: we have different understanding of what is Decomposition(i.e. rootNode & tuple)

File: sc-web/components/scgg/src/scgg-struct.js (hunk 10 above)
Comment (PtaxaMagic) [reply sc-web/components/scgg/src/scgg-struct.js]: that's what i talking about. 
different understanding

File: sc-web/components/scgg/src/scgg-struct.js (hunk 21 above)
Comment (PtaxaMagic) [reply sc-web/components/scgg/src/scgg-struct.js]: rrrrrrr

File: sc-web/components/scgg/src/scgg-struct.js (hunk 7 above)
Comment (blrB) [reply sc-web/components/scgg/src/scgg-struct.js]: if we open graph from struct and graph is have decomposition and mainGraphNode, variable sandbox.graphNodeAddr and sandbox.decompositionNodeAddr will be contain addr

**Pull request merged**